partiql-eval/src/eval.rs

```diff
@@ -46,56 +44,172 @@ pub enum EvaluationError {
     InvalidEvaluationPlan(String),
 }
 
+pub trait Evaluable: Debug {
+    fn evaluate(&mut self, ctx: &dyn EvalContext) -> Option<Value>;
+    fn update_input(&mut self, input: &Value);
+}
+
 #[derive(Debug)]
-pub struct Scan {
+pub struct EvalScan {
     pub expr: Box<dyn EvalExpr>,
     pub as_key: String,
+    pub at_key: String,
     pub output: Option<Value>,
 }
 
-impl Scan {
-    pub fn new(expr: Box<dyn EvalExpr>, as_key: &str) -> Self {
-        Scan {
+impl EvalScan {
+    pub fn new(expr: Box<dyn EvalExpr>, as_key: &str, at_key: &str) -> Self {
+        EvalScan {
             expr,
             as_key: as_key.to_string(),
+            at_key: at_key.to_string(),
             output: None,
         }
     }
 }
 
-impl DagEvaluable for Scan {
+impl Evaluable for EvalScan {
     fn evaluate(&mut self, ctx: &dyn EvalContext) -> Option<Value> {
-        let mut value = partiql_bag!();
+        let mut value = partiql_bag![];
         let v = self.expr.evaluate(&Tuple(HashMap::new()), ctx);
+        let ordered = &v.is_ordered();
+        let mut c: i64 = 0;
+        let at_key = &self.at_key;
         for t in v.into_iter() {
-            let out = Tuple(HashMap::from([(self.as_key.clone(), t)]));
-            value.push(Value::Tuple(Box::new(out)));
+            let mut out = HashMap::from([(self.as_key.clone(), t)]);
+            if !at_key.is_empty() {
+                let at_id = if *ordered { c.into() } else { Missing };
+                out.insert(at_key.clone(), at_id.into());
+                c += 1;
+            }
+
+            value.push(Value::Tuple(Box::new(Tuple(out))));
         }
-        Some(Value::Bag(Box::new(value)))
+
+        self.output = Some(Value::Bag(Box::new(value)));
+        self.output.clone()
     }
+
     fn update_input(&mut self, _input: &Value) {
         todo!("update_input for Scan")
```

Comment:
Just curious, in what scenario would `update_input` be called on an `EvalScan` node?

Reply:
Referring to what I've mentioned in the other comment, it could be the case of last operator from a sub-query to update the `Scan`'s input, but I think we still need to flesh this out.

Reply:
Ok, so like in the example you gave, a subquery within a `FROM` source?

Reply:
Yes.